README.md

docolav
=======

Tous droits (C) 2012-2016 par Mickaël S.

docolav est livré avec ABSOLUMENT AUCUNE GARANTIE.
C'est un programme libre, et vous pouvez le redistribuer sous certaines
conditions. Voir la GNU General Public Licence pour plus de détails.

docolav détermine la couleur moyenne dominante d'une image.


Dépendance
----------

ImageMagick


Installation
------------

	make
	sudo make install


Usage
-----

docolav FORMAT FICHIER

Où FORMAT est une chaîne de caractères passée à printf().
3 arguments sont valables, dans l'ordre : R, G puis B.

Voici quelques exemples :

	docolav "R=%d G=%d B=%d" image.jpg
	docolav "HEX(%02X,%02X,%02X)" image.jpg


CHANGELOG
=== Legend ===
	+ : added
	- : deleted
	^ : bug fixed
	* : optimization
	. : information


*** 0.3.3 (20160110)
	* utilisation totale du C++
	* options de compilation -march=native et -std=c++11
	- test/image.dcm
	. MàJ de CImg.h (1.5.5 rev 1.1232 => 1.6.9)


*** 0.3.2 (20130311)
	^ gestion des arguments
	^ options de compilation pour Windows (sous XP, tout bloquait)
	+ message de CImg en cas d'erreur lors du traitement d'une image
	. MàJ de CImg.h (1.5.3 rev 1.1180 => 1.5.5 rev 1.1232)


*** 0.3.1 (20121025)
	- suppression de l'option de sortie "dim" (restons concentré sur l'objectif principal !)
	^ corrections diverses dans docolav.cc
	* modification du comportement : docolav FORMAT FICHIER (voir README.md)
	+ ajout du Makefile pour Windows
	+ ajout du binaire pour Windows 32 et 64 bits
	. MàJ de CImg.h (1.5.2 => 1.5.3 rev 1.1180)


*** 0.3 (20121024)
	* suppression de la dépendance à OpenCV pour CImg
	+ ajout de l'option de sortie "dim" pour afficher les dimensions de l'image
	+ ajout du binaire pour GNU/Linux 32 bits
	+ ajout des sommes de contrôle des binaires
	^ corrections diverses dans le fichier Makefile


*** 0.2 (20120806)
	+ traduction en français
	+ ajout du binaire pour GNU/Linux 64 bits
	^ corrections diverses dans le fichier Makefile


*** 0.1 (20120206)
	+ ajout de l'option --output-format
	- suppression de l'option --quiet


*** 0.1a (20120131)
	+ premier jet


test/test.sh
#!/bin/sh

super_test() {
	echo -n "\"docolav '%d-%d-%d' $1\" [attendu : $2] => "
	couleurs=$(../docolav '%d-%d-%d' $1)
	echo -n "$couleurs : "
	if [ "$couleurs" = "$2" ]; then
		echo "\033[32mOk\033[0m"
	else
		echo "\033[31mKo\033[0m"
	fi
}

echo "Formats standards"
super_test image.png 161-174-73
super_test image.jpg 161-173-72
super_test image.bmp 161-174-73
super_test image.tif 161-173-72
super_test image.gif 159-174-70
super_test image.ppm 160-173-72
super_test image.tga 161-174-73

echo
echo "Formats exotiques"
super_test image.ico 161-174-73
super_test image.xpm 160-173-72
super_test image.pnm 160-173-72
super_test image.pcx 160-173-72
super_test image.fit 160-160-160
super_test image.pbm 187-187-187
super_test image.pgm 163-163-163
super_test image.xbm 184-184-184


src/docolav.cc

/* =============================================================== */
/*  $Author: Mickaël S. <mickael@jmsinfo.co> $                     */
/*  $Id: docolav $                                                 */
/*  $Revision: 11 $, $Date: 2016/01/10 $                           */
/*  $Source: http://jmsinfo.co $, $Date: 2012/01/31 $              */
/*                                                                 */
/*  docolav détermine la couleur moyenne dominante d'une image.    */
/* =============================================================== */


#include "./docolav.h"
using namespace cimg_library;


// C'est parti mon kiki !
int main(int argc, char *argv[]) {
	// Traitement des arguments
	if ( !argv[1] ) {
		usage();
		return 0;
	}
	std::string format = argv[1];
	if ( format == "-h" || format == "--help" ) {
		help();
		return 0;
	} else if ( format == "-v" || format == "--version" ) {
		version();
		return 0;
	}
	if ( !argv[2] ) {
		usage();
		return 0;
	}
	std::string filename = argv[2];

	// Chargement de l'image et récupération des infos
	try {
		CImg<uint8_t> picture(filename.c_str());
		// Récupération des pixels
		uintptr_t r = 0, g = 0, b = 0;
		if ( picture.spectrum() == 1 ) {
			cimg_forXY ( picture, x, y ) {
				uint8_t *t = picture.data(x, y, 0, 0);
				r += t[0];
				g += t[1];
				b += t[2];
			}
		} else {
			cimg_forXY ( picture, x, y ) {
				r += picture(x, y, 0, 0);
				g += picture(x, y, 0, 1);
				b += picture(x, y, 0, 2);
			}
		}
		// This is it!
		uintptr_t total = picture.width() * picture.height();
		r /= total;
		g /= total;
		b /= total;
		printf(format.c_str(), r, g, b);
		return 0;
	} catch ( CImgIOException &e ) {
		std::cerr << "Impossible de charger l'image : " << filename << std::endl;
		std::cerr << e.what() << std::endl;
		return 1;
	}
}

void help() {
	std::cout <<
	"docolav détermine la couleur moyenne dominante d'une image.\n\n"
	"Usage : docolav FORMAT FICHIER\n\n"
	"FORMAT est une chaîne de caractères passée à printf().\n"
	"3 arguments sont valables, dans l'ordre : R, G puis B.\n\n"
	"Options :\n"
	" -h,  --help      affiche ce message\n"
	" -v,  --version   affiche les versions du programme et de CImg\n\n"
	"Exemples :\n"
	"$ docolav 'R=%d G=%d B=%d' image.jpg\n"
	"R=25 G=52 B=77\n\n"
	"$ docolav 'HEX(%02X,%02X,%02X)' image.png\n"
	"HEX(19,34,4D)\n";
}

void usage() {
	std::cout << "Usage : docolav FORMAT FICHIER" << std::endl;
}

void version() {
	std::cout << "docolav " << DOCOLAV_VERSION << "\n   "
	cimg_appname << " " << cimg_version/100 << "." << (cimg_version/10)%10 << "." << cimg_version%10 << "\n\n"
	"Copyright (C) 2012-2016 by Mickaël S.\n"
	"docolav comes with ABSOLUTELY NO WARRANTY.\n"
	"This is free software, and you are welcome to redistribute it under\n"
	"certain conditions. See the GNU General Public Licence for details.\n\n"
	"Écrit par Mickaël S. <mickael@jmsinfo.co>.\n";
}


src/main.cpp

/* =============================================================== */
/*  $Author: BoboTiG <bobotig@gmail.com> $                         */
/*  $Id: docolav $                                                 */
/*  $Revision: 2 $, $Date: 2012/08/06 $                            */
/*  $Source: http://www.bobotig.fr $, $Date: 2012/01/31 $          */
/*                                                                 */
/*  docolav détermine la couleur moyenne dominante d'une image.    */
/* =============================================================== */


#include <cstdio>
#include <iostream>
#include <string>
#include <opencv/highgui.h>
using namespace std;


// Constantes
#define VERSION "0.2"


// Prototypes
void help();
void usage();
void version();


// C'est parti mon kiki !
int main(int argc, char *argv[]) {
	IplImage* picture = 0;
	unsigned int of = 0, height = 0, width = 0, total = 0;
	unsigned int r = 0, g = 0, b = 0;
	string arg1 = "", img = "", format = "rgb";
	
	if ( argc < 2 ) {
		usage();
		return EXIT_FAILURE;
	}
	arg1 = argv[1];
	if ( arg1 == "-h" || arg1 == "--help" ) {
		help();
		return EXIT_SUCCESS;
	} else if ( arg1 == "-v" || arg1 == "--version" ) {
		version();
		return EXIT_SUCCESS;
	} else if ( arg1 == "-of" || arg1 == "--output-format" ) {
		of = 1;
		if ( argc != 4 ) {
			usage();
			return EXIT_FAILURE;
		}
		format = argv[2];
		if ( format != "rgb" && format != "hex" && format != "html" ) {
			cout << "Formats disponibles : rgb, hex et html." << endl;
			return EXIT_FAILURE;
		}
		img = argv[3];
	} else {
		img = arg1;
	}
	
	// Chargement de l'image
	picture = cvLoadImage(img.c_str(), CV_LOAD_IMAGE_UNCHANGED);
	if ( ! picture ) {
		cout << "Impossible de charger l'image : " << img << endl;
		return EXIT_FAILURE;
	}
	
	// Affichage de la bête
	/*cvNamedWindow("une_fentere");
	cvShowImage("une_fentere", picture);
	cvWaitKey();
	cvDestroyWindow("une_fentere");
	return 0;
	//*/
	
	// Récupération des tous les pixels
	height = picture->height;
	width  = picture->width;
	for ( unsigned int x = 0; x < width; ++x ) {
		for ( unsigned int y = 0; y < height; ++y ) {
			uchar* ptr = cvPtr2D(picture, y, x, NULL);
			r += (int) ptr[2];
			g += (int) ptr[1];
			b += (int) ptr[0];
			++total;
		}
	}
	cvReleaseImage(&picture);
	r /= total;
	g /= total;
	b /= total;
	
	// This is it !
	if ( of ) {
		if      ( format == "rgb" ) printf("RGB(%d,%d,%d)", r, g, b);
		else if ( format == "hex" ) printf("HEX(%02X,%02X,%02X)", r, g, b);
		else                        printf("#%02X%02X%02X", r, g, b);
	} else {
		cout << "Fichier    : " << img << endl;
		cout << "Dimensions : " << width << " * " << height << " px" << endl;
		cout << "Couleur moyenne dominante" << endl;
		printf("    RGB(%d,%d,%d)\n", r, g, b);
		printf("    HEX(%02X,%02X,%02X)\n", r, g, b);
		printf("    #%02X%02X%02X\n", r, g, b);
	}
	return EXIT_SUCCESS;
}


void help() {
	cout << "Copyright (C) 2012 by BoboTiG.\n\
\n\
docolav comes with ABSOLUTELY NO WARRANTY.\n\
This is free software, and you are welcome to redistribute it under \n\
certain conditions. See the GNU General Public Licence for details.\n\
\n\
docolav détermine la couleur moyenne dominante d'une image.\n\
\n\
Usage : docolav [OPTION [VALEUR]] FICHIER\n\
\n\
Options :\n\
 -h,  --help		affiche ce message\n\
 -of, --output-format	format de sortie : rgb, hex ou html\n\
 -v,  --version		affiche les versions du programme et d'OpenCV\n\
\n\
Rapporter erreurs, suggestions et commentaires à <bobotig@gmail.com>." << endl;
}


void usage() {
	cout << "Usage : docolav [OPTION [VALEUR]] FICHIER" << endl;
}


void version() {
	cout << "docolav v" << VERSION << "." << endl;
	cout << "OpenCV v" << CV_VERSION << "." << endl;
}
